vm/core/context.go

```diff
@@ -112,74 +218,56 @@ func (c *Context) transfer(from *Account, to Address, amount, max uint64) error
 	if amount > from.Balance {
 		return ErrNoBalance
 	}
-	if c.transferred+amount > max {
+	if totalTransfer, err := safeAdd(c.transferred, amount); err != nil {
+		return err
+	} else if totalTransfer > max {
 		return fmt.Errorf("%w: %d", ErrMaxSpend, max)
 	}
+
 	// noop. only gas is consumed
 	if from.Address == to {
 		return nil
 	}
 
+	if newBalance, err := safeAdd(account.Balance, amount); err != nil {
+		return err
+	} else {
+		account.Balance = newBalance
+	}
 	c.transferred += amount
 	from.Balance -= amount
-	account.Balance += amount
+	c.change(from)
 	c.change(account)
+	c.Logger.Debug(
+		"transfer",
+		zap.Uint64("amount", amount),
+		zap.String("from", from.Address.String()),
+		zap.String("to", to.String()),
```

Comment:
Also here and in few other places:
```suggestion
		zap.Stringer("from", from.Address),
		zap.Stringer("to", to),
```